internal/usde/usde.go

```diff
@@ -1,39 +1,98 @@
+// It returns the strategy to use and a boolean pointer indicating if the specs are different (if the specs were not compared, then nil will be returned).
+func DeriveUpgradeStrategy(ctx context.Context, newSpec *kubernetes.GenericObject, existingSpec *kubernetes.GenericObject,
+	inProgressUpgradeStrategy string, overrideToPPND, overrideToProgressive *bool) (UpgradeStrategy, *bool, error) {
+
+	numaLogger := logger.FromContext(ctx)
+
+	if UpgradeStrategy(inProgressUpgradeStrategy) == UpgradeStrategyPPND || (overrideToPPND != nil && *overrideToPPND) {
+		return UpgradeStrategyPPND, nil, nil
+	}
 
-func (s *USDEStrategy) UnmarshalJSON(data []byte) (err error) {
-	// Trim spaces and check length
-	dataStrNoSpaces := strings.TrimSpace(string(data[:]))
-	if len(dataStrNoSpaces) == 0 {
-		return fmt.Errorf("empty strategy (allowed values are: %s or %s)", progressiveStrategyID, ppndStrategyID)
+	if UpgradeStrategy(inProgressUpgradeStrategy) == UpgradeStrategyProgressive || (overrideToProgressive != nil && *overrideToProgressive) {
+		// TODO-PROGRESSIVE: return UpgradeStrategyProgressive instead of UpgradeStrategyPPND
+		return UpgradeStrategyPPND, nil, nil
 	}
 
-	// Remove the double quotes around the string
-	strategyStr := string(data[1 : len(data)-1])
+	// Get USDE Config
+	usdeConfig := config.GetConfigManagerInstance().GetUSDEConfig()
+
+	// Get apply paths based on the spec type (Pipeline, ISBS)
+	applyPaths := []string{}
+	if reflect.DeepEqual(newSpec.GroupVersionKind(), numaflowv1.PipelineGroupVersionKind) {
+		applyPaths = usdeConfig.PipelineSpecExcludedPaths
+	} else if reflect.DeepEqual(newSpec.GroupVersionKind(), numaflowv1.ISBGroupVersionKind) {
+		applyPaths = usdeConfig.ISBServiceSpecExcludedPaths
+	}
+
+	numaLogger.WithValues("usdeConfig", usdeConfig, "applyPaths", applyPaths).Debug("started deriving upgrade strategy")
+
+	// Split newSpec
+	newSpecOnlyApplyPaths, newSpecWithoutApplyPaths, err := util.SplitObject(newSpec.Spec.Raw, applyPaths, ".")
+	if err != nil {
+		return UpgradeStrategyError, nil, err
+	}
+
+	numaLogger.WithValues("newSpecOnlyApplyPaths", newSpecOnlyApplyPaths, "newSpecWithoutApplyPaths", newSpecWithoutApplyPaths).Debug("split new spec")
+
+	// Split existingSpec
+	existingSpecOnlyApplyPaths, existingSpecWithoutApplyPaths, err := util.SplitObject(existingSpec.Spec.Raw, applyPaths, ".")
+	if err != nil {
+		return UpgradeStrategyError, nil, err
+	}
+
+	numaLogger.WithValues("existingSpecOnlyApplyPaths", existingSpecOnlyApplyPaths, "existingSpecWithoutApplyPaths", existingSpecWithoutApplyPaths).Debug("split existing spec")
+
+	specsDiffer := true
+
+	// Compare specs without the apply fields and check user's strategy to either return PPND or Progressive
+	if !reflect.DeepEqual(newSpecWithoutApplyPaths, existingSpecWithoutApplyPaths) {
+		userUpgradeStrategy := config.GetConfigManagerInstance().GetNamespaceConfig(newSpec.Namespace).UpgradeStrategy
```

Comment:
is it possible for the `GetNamespaceConfig()` function to fail if there's no config for the namespace? Can we have some default value if that file doesn't exist?